[Pytorch][ATEN] Enable FP8 concatenate

jwfromm opened the pull request:
Summary: Float8 is becoming and increasingly popular datatype now that it is well supported on GPUs. This  diff enables FP8 to work with `torch.cat`. This is pretty straight forward since memory operations dont vary based on the input dtype, but can be quite helpful for FP8 based models.

Test Plan:
```
buck2 run mode/opt -c fbcode.enable_gpu_sections=true -c fbcode.platform=platform010 -c fbcode.nvcc_arch=h100a -c fbcode.platform010_cuda_version=12 //caffe2/test:tensor_creation -- -r test_cat_all_dtypes_and_devices
```

Differential Revision: D64443965




cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

qchip reviewed (approved):
thanks @jwfromm for adding fp8 support to torch.cat.

Skylion007 commented on aten/src/ATen/native/cpu/CatKernel.cpp:
```diff
@@ -55,14 +60,24 @@ void cat_serial_kernel_impl(const Tensor& result, const MaterializedITensorListR
+    const Tensor& result,
+    const MaterializedITensorListRef& tensors,
+    int64_t dim) {
+  AT_DISPATCH_FLOATING_TYPES_AND6(
```
Use AT_DISPATCH_V2. Makes your life so much easier here `AT_EXPAND(FLOAT8_DTYPES)`

jwfromm replied:
This is much nicer, thank you for the tip! The latest commit looks pretty clean.

Skylion007 commented on aten/src/ATen/native/cuda/Shape.cu:
```diff
@@ -518,10 +532,27 @@ TORCH_IMPL_FUNC(cat_out_cuda)
+            kHalf,
+            kBool,
+            kBFloat16,
+            kFloat8_e4m3fn,
+            kFloat8_e4m3fnuz,
+            kFloat8_e5m2,
+            kFloat8_e5m2fnuz,
```
```suggestion
            AT_EXPAND(AT_FLOAT8_TYPES),
```

Skylion007 commented on aten/src/ATen/native/cuda/Shape.cu:
```diff
@@ -500,10 +500,24 @@ TORCH_IMPL_FUNC(cat_out_cuda)
+            kHalf,
+            kBool,
+            kBFloat16,
+            kFloat8_e4m3fn,
+            kFloat8_e4m3fnuz,
+            kFloat8_e5m2,
+            kFloat8_e5m2fnuz,
```
```suggestion
            AT_EXPAND(AT_FLOAT8_TYPES),
```

Skylion007 commented:
After these changes, you won't need the new macros, either.

jwfromm committed:
[Pytorch][ATEN] Enable FP8 concatenate (#138046)

Summary:

Float8 is becoming and increasingly popular datatype now that it is well supported on GPUs. This  diff enables FP8 to work with `torch.cat`. This is pretty straight forward since memory operations dont vary based on the input dtype, but can be quite helpful for FP8 based models.

Test Plan:
```
buck2 run mode/opt -c fbcode.enable_gpu_sections=true -c fbcode.platform=platform010 -c fbcode.nvcc_arch=h100a -c fbcode.platform010_cuda_version=12 //caffe2/test:tensor_creation -- -r test_cat_all_dtypes_and_devices
```

Reviewed By: qchip

Differential Revision: D64443965
